spark_log_parser/loaders/__init__.py

```diff
@@ -61,58 +71,110 @@ def __iter__(self):
                 self._add_chunk_to_size(chunk)
                 yield chunk
 
-    def _add_chunk_to_size(self, chunk):
+    def _add_chunk_to_size(self, chunk) -> None:
         self.total_size += len(chunk)
-        if self._maximum_allowed_size is not None and self.total_size > self._maximum_allowed_size:
+        if self.total_size > self._maximum_allowed_size:
             raise ArchiveTooLargeError()
 
-    def read(self, size=-1) -> bytes:
-        content = self._trailing_data
-        self._trailing_data = b''
+    def _num_bytes_left_in_buffer(self) -> int:
+        return len(self._buffer) - self._buffer_position
 
-        if not size or size == -1:
-            for chunk in self._chunks:
-                self._add_chunk_to_size(chunk)
-                content += chunk
-        else:
-            # iterate chunks until we've read at least `size` amount of data
+    def read(self, size=-1) -> bytes | bytearray | memoryview:
```

Comment:
Seems like we could refactor this function to always return a memoryview (though I think we'd still want the return type to by bytes). Replace the outer if statement with something like,
```
if size < 0:
  size = None

while not size or self._num_bytes_left_in_buffer() < size:
  try:
    chunk = next(self._chunks)
    self._add_chunk_to_size(chunk)
    self._buffer += chunk
  except StopIteration:
    break
  start_pos = self._buffer_position
  self._buffer_position = self._buffer_position + size if size else len(self._buffer)
  return self._buffer_memview[start_pos:self._buffer_position]
```
That can be improved on and I'm probably even missing something, but if this works then we're always returning the same thing and I think it's easier to understand.

Reply:
> Do you think it makes sense to just have bytes here since the return value will always have that protocol?

If we return a `memoryview`, this isn't actually true - `memoryview`s don't support `splitlines()`, for example, which is why I have to do this `bytearray` copy in `iter_lines` - https://github.com/synccomputingcode/spark_log_parser/pull/39/files#diff-e6eadcc243cd6e0abd0b257ed701c0e4b5a24d8712a487a823c9f3a929f561c3R150

In general, the `memoryview` API is quite bare-bones compared to `bytes`/`bytearray` - https://docs.python.org/3/library/stdtypes.html#memory-views. They do play nicely with operators like `+`, though, or slice assignments (i.e. `some_buffer[:index] = my_memview`)

I do agree with you though that it would be best to settle on 1 return type - if we keep the memview, then I would vote for just `read() -> memoryview`, otherwise `bytearray`

Reply:
Makes sense to land on the type descriptor that's fully supported by the returned object in all cases. In that case, it matters less that the actual type of the returned object is conditional. I'd feel better if the actual type of the returned object were uniform, but happy to approve if that's not easily achieved/less performant or otherwise doesn't make sense.